src/elm/Pages/Build/Graph/DOT.elm

```diff
@@ -0,0 +1,505 @@
+
+        pipelineEdges =
+            edges
+                |> List.filter (\e -> e.label.cluster == pipelineClusterID)
+
+        serviceEdges =
+            edges
+                |> List.filter (\e -> e.label.cluster == serviceClusterID)
+
+        -- convert nodes and edges to DOT string format
+        builtInNodesString =
+            List.map (nodeToString model repo build) builtInNodes
+                |> String.join "\n"
+
+        pipelineNodesString =
+            List.map (nodeToString model repo build) pipelineNodes
+                |> String.join "\n"
+
+        serviceNodesString =
+            List.map (nodeToString model repo build) serviceNodes
+                |> String.join "\n"
+
+        builtInEdgesString =
+            List.map edgeToString builtInEdges
+                |> String.join "\n"
+
+        pipelineEdgesString =
+            List.map edgeToString pipelineEdges
+                |> String.join "\n"
+
+        serviceEdgesString =
+            List.map edgeToString serviceEdges
+                |> String.join "\n"
+
+        -- construct DOT subgraphs using nodes and edges
+        pipelineSubgraph =
+            clusterSubgraph pipelineClusterID pipelineSubgraphStyles pipelineNodesString pipelineEdgesString
+
+        builtInSubgraph =
+            clusterSubgraph builtInClusterID builtInSubgraphStyles builtInNodesString builtInEdgesString
+
+        serviceSubgraph =
+            if model.repo.build.graph.showServices then
+                clusterSubgraph serviceClusterID serviceSubgraphStyles serviceNodesString serviceEdgesString
+
+            else
+                ""
+
+        -- reverse the subgraphs for top-bottom rankdir to consistently group services and built-ins
+        rotation =
+            case model.repo.build.graph.rankdir of
+                TB ->
+                    List.reverse
+
+                _ ->
+                    identity
+
+        subgraphs =
+            [ -- pipeline (stages, steps) subgraph and cluster
+              pipelineSubgraph
+            , ""
+
+            -- built-in (init, clone) subgraph and cluster
+            , builtInSubgraph
+            , ""
+
+            -- services subgraph and cluster
+            , serviceSubgraph
+            ]
+    in
+    digraph (baseGraphStyles model.repo.build.graph.rankdir)
+        (rotation subgraphs)
+
+
+{-| nodeLabel : takes model, graph info, a node, and returns a string representation of the "label" applied to a node element.
+a "label" is actually a disguised graphviz table <https://graphviz.org/doc/info/lang.html> that is used to
+render a list of stage-steps as graph content that is recognized by the layout
+-}
+nodeLabel : BuildModel.PartialModel a -> Repository -> Build -> BuildGraphNode -> Bool -> String
+nodeLabel model repo build node showSteps =
+    let
+        label =
+            node.name
+
+        steps =
+            List.sortBy .id node.steps
+
+        table content =
+            "<table "
```

Comment:
probably not much we can do here, but too bad we have to string concatenate html 😢

Reply:
i mean i 100% agree the code screams "hacky" 

im open to alternatives, but its not _really_ html, its "html-like" strings, following the graphviz docs:
https://graphviz.org/docs/attrs/label/

Reply:
yup, no worries. necessary evil